### ISSUE TITLE
Lightmapper raytrace support not built into M1 native binary

### ISSUE THREAD

**Issue by ViktorEvil** [godotengine/godot#48082]
**Godot version:**
3.3.stable.official

**OS/device including version:**
macOS 11.2.3 on an M1 MacBook Pro

**Issue description:**
Lightmapper does not work and displays error message "Godot editor was built without ray tracing support, lightmaps can't be baked."

**Steps to reproduce:**
Clicking the "Bake Lightmaps" button results in the error message.
However running the Godot binary in Rosetta mode results in the Lightmaps being baked but obviously this is running non-natively 

**Minimal reproduction project:**
[LightBakeNew.zip](https://github.com/godotengine/godot/files/6357133/LightBakeNew.zip)
project downloaded from reddit

**Comment by bruvzg** [comment]
It's built without lightmapper since version OIDN library used in it have no ARM support. Upstream version have ARM64 support added, but there are some issues with updating, see #47344.

**Comment by akien-mga** [comment]
It's actually not just OIDN (the lightmapper could still be used without denoiser), but the main issue is the Embree library is x86_64 only.

There's a fork adding aarch64 support but we haven't got around to testing it in time for 3.3. This should be work on eventually.

https://github.com/lighttransport/embree-aarch64

Marking this as "enhancement" as it's not something we can fix in Godot code per se as the thirdparty library explicitly doesn't support aarch64. But hopefully the fork will work well.

As a workaround, you should be able to force using the x86_64 version with Rosetta 2 to bake your lightmaps.

**Comment by ViktorEvil** [comment]
> As a workaround, you should be able to force using the x86_64 version with Rosetta 2 to bake your lightmaps.

Yes Rosetta works fine so its not a massive issue.

**Comment by JFonS** [comment]
Yeah, once I'm done getting embree-aarch64 to build on master it should be easy to back-port to 3.x.

**Comment by jasonliu--** [comment]
It looks like the Blender folks are using [sse2neon.h](https://github.com/DLTcollab/sse2neon) in order to allow them to compile Blender on arm64 with OIDN and Embree included.

See also:
* https://github.com/DLTcollab/sse2neon#adoptions
* https://developer.blender.org/T78710
* https://developer.blender.org/rBL62566

I happen to stumble across this issue because I am trying to [package](https://ports.macports.org/port/godot/) Godot 3.3 for [MacPorts](https://github.com/macports/macports-ports).

**Comment by jasonliu--** [comment]
It was also brought up on the MacPorts dev mailing list yesterday that in addition to sse2neon, there is another header-only library called SIMDe (SIMD everywhere), which pretty much does the same thing. Either of these libraries might be a good choice to allow Godot to simply use the "official" Embree and OIDN libraries on arm64. (And probably other libraries as well, as Godot continues to grow and expand.)

https://github.com/simd-everywhere/simde

**Comment by akien-mga** [comment]
@ViktorEvil @jasonliu-- Could you test this build to see if the lightmapper now works as expected on Apple M1 with the native aarch64 binary? https://downloads.tuxfamily.org/godotengine/testing/3.3.1-rc-476bc5191/

This is a build from the latest [`3.3` branch](https://github.com/godotengine/godot/tree/3.3) (upcoming 3.3.1). @jasonliu-- You might want to check if that improves things for MacPorts packaging.

Note that the denoiser (`modules/denoise`) is still not expected to build on aarch64 as the older OIDN version we use is x86_64 only. We'll solve that eventually, either by upgrading (but then it requires ispc, which is a pain) or by patching it support ourselves.

**Comment by ViktorEvil** [comment]
> Note that the denoiser (`modules/denoise`) is still not expected to build on aarch64 as the older OIDN version we use is x86_64 only. We'll solve that eventually, either by upgrading (but then it requires ispc, which is a pain) or by patching it support ourselves.

Raytracing now works great with just the denoiser that you mentioned that is not working. Thanks for the progress on this, I as well as other Apple Silicon users appreciate the engine devs not leaving minority users like myself out. I know it would be a lot easier to just build for x86.

**Comment by jasonliu--** [comment]
I don't personally own an M1 Mac, so I'll need to get one of the other MacPorts devs to try the build.

> Note that the denoiser (modules/denoise) is still not expected to build on aarch64 as the older OIDN version we use is x86_64 only. We'll solve that eventually, either by upgrading (but then it requires ispc, which is a pain) or by patching it support ourselves.

This is why I think using `sse2neon` or `SIMDe` is probably a better solution in the long term than using `embree-aarch64`. `sse2neon` and `SIMDe` allow you to run code that uses Intel intrinsics on ARM processors. It's nice that someone went through the effort of porting Embree to ARM, but there are plenty of other libraries (such as OIDN), which haven't been ported. `sse2neon` and `SIMDe` allow you to simply run those original x86_64 codes on ARM, with very minimal patching:

```
// This patch would be made to Embree's or OIDN's source code
#ifdef __arm___ // <-- This is just an example, and probably not what you would actually use
  #include <sse2neon.h>
#else
  #include <emmintrin.h> // <-- This is the header for using Intel SSE2
#endif
```

In fact, the MacPorts devs think that these libraries are so useful, that I am preparing packages for both `sse2neon` and `SIMDe` to be added to MacPorts, so that our package maintainers can start adding ARM support for libraries which haven't yet been rewritten to support ARM.

**Comment by akien-mga** [comment]
AFAIK embree-aarch64 does use sse2neon. You can check https://github.com/godotengine/godot/pull/48455 to see the diff between Intel embree and the fork with aarch64 support. It's not just SSE but also stuff like wrong arch specific assumptions on primitives types.

**Comment by jasonliu--** [comment]
Yes, I think you are correct. It looks like embree-aarch64 is using the original SSE2NEON that was written by John Ratcliff. In addition, it looks like embree-aarch64 is implementing its own header to convert AVX to Neon. `sse2neon` (both Ratcliff's original and the current one) don't provide support for translating AVX to Neon... which would seem to imply that `SIMDe` might actually be a better solution. (`SIMDe` is essentially a superset of `sse2neon`, and actually includes `sse2neon`, as well as support for AVX, and even WebAssembly's SIMD128.)

**Comment by jasonliu--** [comment]
One more thing... AFAICT, SIMDe apparently also works in translating in the other direction as well, which means that it has headers that allow code which has been written for ARM Neon to be run on x86 processors.

Sorry if it seems like I'm pushing especially hard for using sse2neon/SIMDe.... it's just that it seems like a more generalized solution to the SIMD intrinsics issue, and not just a solution for one library.

**Comment by akien-mga** [comment]
Closing as fixed by #48455.

For porting OIDN to ARM64, let's focus the discussion on #47344 (either upgrading - and adding ISPC as a build dependency - or trying to patch things with sse2non or SIMDe).
